### PR TITLE
parse_rule: skip rule definition binder creation

### LINKED ISSUE #643: Why not rm the SIOF by using template metaclass specialized on rule_id?
Seth proposed a solution to a problem appearing on stackoverflow here:

  https://github.com/sehe/splitting-large-x3-parsers

Another solution was proposed here:

  https://github.com/sehe/splitting-large-x3-parsers/issues/1

which eliminated the SIOF problem with Seth's solution. 
Since Seth hasn't responded to that proposed solution, I was
hoping someone else might; hence, I'm announcing it here.
Note, this proposed solution would eliminate the need for
the rule_definition; hence, would break any existing spirit-x3
code that uses it, but that might be avoided using some sort
of #pragma.

### PR DESCRIPTION
**Pull request by Kojoley**
Not all of our parsers are constexpr constructible, and static local variable is not free for these.

Also, closes #643.

### CODE REVIEW

**Review by cppljevans — CHANGES_REQUESTED** [review]
In short, you're rm'ing the 'static auto const def' and instead inlining that def.parse.  However, not quite, AFAICT.  Shouldn't  BOOST_JOIN(rule_name, _def).rhs be the 1st arg to call_rule_definition
in order to get the RHS of the rule?

**Review by cppljevans — COMMENTED** [review]
Why not just eliminate the static and use the rule_definition:parse directly instead of inlining it?
IOW, just this one line:

  return BOOST_JOIN(rule_name, _def).parse(first, last, context, unused, attr);

**Comment by Kojoley** [comment]
> Shouldn't BOOST_JOIN(rule_name, _def).rhs be the 1st arg to call_rule_definition

No

> Why not just eliminate the static and use the rule_definition:parse directly instead of inlining it?

Because a temporary is even worse than the static local.

> IOW, just this one line:
> 
> return BOOST_JOIN(rule_name, _def).parse(first, last, context, unused, attr);

This will break annotations and debug handlers.

**Comment by djowel** [comment]
@Kojoley  'Some checks were not successful", can you please check? I'd like to merge this.

**Comment by cppljevans** [comment]
> 
> 
> > Shouldn't BOOST_JOIN(rule_name, _def).rhs be the 1st arg to call_rule_definition
> 
> No
> 
> > Why not just eliminate the static and use the rule_definition:parse directly instead of inlining it?
> 
> Because a temporary is even worse than the static local.
> 
> > IOW, just this one line:
> > return BOOST_JOIN(rule_name, _def).parse(first, last, context, unused, attr);
> 
> This will break annotations and debug handlers.

Ah, I see my mistake.  I had assumed, based on previous views of rule_definition use, that the
BOOST_JOIN(rule_name,_def) was an instance of rule_definition.  However, now
looking at all the examples and test (in particular, the https://github.com/Kojoley/spirit/blob/parse_rule-optimization/test/x3/rule_separate_tu_grammar.cpp), that none of the _def names are rule_definitions.
Sorry for the mistake, and of course that mistake also explains why I thought BOOST_JOIN(rule_name,_def)_.parse
would be a call to rule_definition<...>::parse.  I realize now that was a rather foolish assumption :(

Sorry for noise.

**Comment by Kojoley** [comment]
> @Kojoley 'Some checks were not successful", can you please check? I'd like to merge this.

Boost.Math C++11 requirement broke some of our tests, I was waiting for CI green light to push fixes to develop branch, but Travis turnaround times these days are extreme.

**Pull request merged**
